### ISSUE TITLE
setData performance decreased in safari compared to v1.0.0

### ISSUE THREAD

**Issue by mmuelder** [mapbox/mapbox-gl-js#10187]
Greetings!

In the current version 2.0.0 (as far as I know actually since 1.10) the performance of setData combined with a moving camera has extremely decreased in comparison to 1.0.0. We only encountered this problem in the safari on iphones and ipads, no other browsers seem to have this issue, at least not in this magnitude. The circle layer in this example is super laggy:

### V 1.0.0 (left) versus V 2.0.0 (right)
![mapbox1](https://user-images.githubusercontent.com/5302657/101879780-1fd13380-3b92-11eb-9ac3-54271a0fdb4c.gif)![mapbox2](https://user-images.githubusercontent.com/5302657/101879785-219af700-3b92-11eb-87aa-55a9f356b2fa.gif)
(recorded on safari @ ipad air 2 12.4.1, same problem on all tested iphones)

### Steps to Trigger Behavior

Use setData in combination with a moving camera (easeTo), maybe it also has something to do with circle layers.

### Link to Demonstration

The code of the examples seen above can be found here:
https://jsbin.com/hagivowiko/edit?html,output

just replace 
```
<script src='https://api.mapbox.com/mapbox-gl-js/v2.0.0/mapbox-gl.js'></script>
<link href='https://api.mapbox.com/mapbox-gl-js/v2.0.0/mapbox-gl.css' rel='stylesheet' />
```

with
```
<script src='https://api.mapbox.com/mapbox-gl-js/v1.0.0/mapbox-gl.js'></script>
<link href='https://api.mapbox.com/mapbox-gl-js/v1.0.0/mapbox-gl.css' rel='stylesheet' />
```

to see the increased performance (in safari). 

### Expected Behavior

at least the same performance as 1.0.0

### Actual Behavior

extremely laggy drawing/update of the circle layer

**Comment by stephenheron** [comment]
We are seeing something similar, we were on an old v0.53 version and I upgraded to the latest v2.0 and calls to setData in Safari is a lot slower. 

I went back and checked the versions, for us v1.2.1 was speedy but anything past that, the setData performance is pretty bad.

**Comment by stephenheron** [comment]
I did a little more debugging and I think commit 5f8f1c236cc9967309a6696aa8e3efca871f88ff might be the issue. It is quite a chunky commit so not had time to dig into it yet.

**Comment by stephenheron** [comment]
I had a little bit more digging and I found some potentially interesting findings. 

I had a look at the at the latest v2.0 version and it seems like there something strange going on with the [ThrottledInvoker](https://github.com/mapbox/mapbox-gl-js/blob/main/src/util/throttled_invoker.js) class. If I disable the "MessageChannel" branch and instead use setTimeout the performance is a lot better for me in Safari. 

I did added some timing debugging to the class to see how long the messages were taking. You can see the modified class here: https://gist.github.com/stephenheron/923bad08383995025c1ae184cf8e7c0d (Lines 20 and 32).

Safari Output: 
![image](https://user-images.githubusercontent.com/142855/103095708-7eb69400-45f9-11eb-89cc-79fcc228456e.png)
 
Chrome Output:
![image](https://user-images.githubusercontent.com/142855/103095761-b02f5f80-45f9-11eb-9eb5-51c1b0cfeafd.png)

Firefox Output:
![image](https://user-images.githubusercontent.com/142855/103095833-ecfb5680-45f9-11eb-95b9-0098d0ab360f.png)

I am not quite sure what to make of this as I can't really find anything that suggests why the MessageChannel appears to be so slow in Safari. The difference is so large that it has me questioning my measurements.

**Comment by vitaliyhayda** [comment]
same degraded performance issue in Safari on iOS devices

**Comment by mourner** [comment]
Submitted a PR addressing this issue in #10417 — please try it on your use cases and let me know if there are any problems with it!

> I did a little more debugging and I think commit [5f8f1c2](https://github.com/mapbox/mapbox-gl-js/commit/5f8f1c236cc9967309a6696aa8e3efca871f88ff) might be the issue. It is quite a chunky commit so not had time to dig into it yet.

@stephenheron thanks so much for bisecting to this commit, this made my job fixing this much easier!

**Comment by stephenheron** [comment]
@mourner Nice one. I just tried out your fix and it is 100% better now in Safari for me. Can't tell the difference between Safari and Chrome/Firefox whereas before it was very noticeable. 

Glad my bisecting was useful, I tried to help further but the code in that commit was very complicated. Thanks for the detailed PR so I can see what was changed and hopefully learn what was going on as it was bugging me that I couldn't figure it out.
